src/Draco.Compiler/Syntax/ParseTree.cs

```diff
@@ -0,0 +1,144 @@
+using Draco.Compiler.Utilities;
+
+namespace Draco.Compiler.Syntax;
+
+/// <summary>
+/// An immutable structure representing a parsed source text with information about concrete syntax.
+/// </summary>
+internal abstract record class ParseTree
+{
+    public sealed record class CompilationUnit(
+        ValueArray<Decl> Declarations) : ParseTree;
+
+    public abstract record class Decl : ParseTree
+    {
+        public sealed record class Func(
+            IToken FuncKeyword,
+            IToken Identifier,
+            IToken OpenParenToken,
+            ValueArray<(
+                (   IToken Identifier,
+                    TypeSpecifier Type
+                ) Param,
+                IToken CommaToken
+            )> Params,
+            IToken CloseParenToken,
+            TypeSpecifier? Type,
+            FuncBody Body) : Decl;
+
+        public sealed record class Label(
+            IToken Identifier,
+            IToken ColonToken) : Decl;
+
+        public sealed record class Variable(
+            IToken Keyword, // Either var or val
+            IToken Identifier,
+            TypeSpecifier? Type,
+            (   IToken EqualsToken,
+                Expr Expression
+            )? Initializer) : Decl;
+    }
+
+    public abstract record class FuncBody
+    {
+        public sealed record class BlockBody(
+            Block Block) : FuncBody;
+
+        public sealed record class InlineBody(
+            IToken EqualsToken,
+            Expr Expression) : FuncBody;
+    }
+
+    public abstract record class TypeName : ParseTree
+    {
+        // This is the only kind of type name for now
+        public sealed record class Simple(
+            IToken Identifier) : TypeName;
+    }
+
+    public sealed record class TypeSpecifier(
+        IToken ColonToken,
+        TypeName Type) : ParseTree;
+
+    public sealed record class Block(
+        ValueArray<Stmt> Statements,
+        Expr? Expression) : ParseTree;
+
+    public abstract record class Stmt : ParseTree
+    {
+        public sealed record class DeclStmt(
```

Comment:
Why duplicate `Stmt` in the name?

Reply:
Calling it `Decl` makes the compiler complain that it hides `ParseTree.Decl`, but eh not a big issue.